[Azure Maps] Timezone: Swagger improvement for SDK generation [Hub Generated]

alextts627 opened the pull request:
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/alankashiwa/azure-rest-api-specs/maps-geolocation-improvement-for-sdk...main/).


### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [x] **To improve swagger for generated clients and address some issues from [this review in ApiView](https://apiview.dev/Assemblies/Review/b73971f62137493dbdd7a10b704eee13)**
  2. When are you targeting to deploy the new service/feature to public regions? **Services already deployed.**
  3. When do you expect to publish the swagger? **ASAP**
  4. If updating an existing version, please select the specific language SDKs and CLIs that must be refreshed after the swagger is published. **No data-plane SDKs are onboarded yet, so no refresh is required.**

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further questions about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [x] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

AlexTsao committed:
Rename with `x-ms-client-name`

AlexTsao committed:
Add extensible enum for timezone accept language

AlexTsao committed:
Rename operationId to have better method meaning

Add x-ms-client-name for better explaining

AlexTsao committed:
Updated description for HasZone1970Location property

AlexTsao committed:
Removed Enum in model class name

AlexTsao committed:
Better naming for TimeTransitions

alankashiwa commented on specification/maps/data-plane/Timezone/preview/1.0/timezone.json:
```diff
@@ -264,7 +357,7 @@
-        "operationId": "GetTimezoneEnumIANA",
+        "operationId": "GetTimezoneIANA",
         "x-ms-examples": {
           "GetTimezoneEnumIANA": {
```
Change to a friendly description for all `x-ms-examples`

alankashiwa replied:
Change to a friendly description for all `x-ms-examples`

alankashiwa commented on specification/maps/data-plane/Timezone/preview/1.0/timezone.json:
```diff
@@ -231,7 +324,7 @@
         "description": "__Enumerate Windows Time Zones__\n\n\n**Applies to**: S0 and S1 pricing tiers.\n\n\nThis API returns a full list of Windows Time Zone IDs.",
-        "operationId": "GetTimezoneEnumWindows",
+        "operationId": "GetTimezoneWindows",
         "x-ms-examples": {
```
Change to a friendly description for all `x-ms-examples`

alankashiwa commented on specification/maps/data-plane/Timezone/preview/1.0/timezone.json:
```diff
@@ -330,7 +423,7 @@
-        "operationId": "GetTimezoneWindowsToIANA",
+        "operationId": "GetIANAMetadata",
         "x-ms-examples": {
           "GetTimezoneWindowsToIANA": {
```
Change to a friendly description for all `x-ms-examples`

alankashiwa commented:
Maybe you can rebase upstream/main given that the previous PR is merged and that may cause conflicts?

alankashiwa commented on specification/maps/data-plane/Timezone/preview/1.0/timezone.json:
```diff
@@ -231,7 +324,7 @@
     "/timezone/enumWindows/{format}": {
       "get": {
         "description": "__Enumerate Windows Time Zones__\n\n\n**Applies to**: S0 and S1 pricing tiers.\n\n\nThis API returns a full list of Windows Time Zone IDs.",
```
Modify description as well?

alankashiwa commented on specification/maps/data-plane/Timezone/preview/1.0/timezone.json:
```diff
@@ -264,7 +357,7 @@
     "/timezone/enumIana/{format}": {
       "get": {
         "description": "__Enumerate IANA Time Zones__\n\n\n**Applies to**: S0 and S1 pricing tiers.\n\n\nThis API returns a full list of IANA time zone IDs. Updates to the IANA service will be reflected in the system within one day.",
```
Modify description as well?

alankashiwa commented on specification/maps/data-plane/Timezone/preview/1.0/timezone.json:
```diff
@@ -297,9 +390,9 @@
     "/timezone/ianaVersion/{format}": {
       "get": {
         "description": "__Time Zone IANA Version__\n\n\n**Applies to**: S0 and S1 pricing tiers.\n\n\nThis API returns the current IANA version number.",
```
Modify description as well?

alankashiwa commented on specification/maps/data-plane/Timezone/preview/1.0/timezone.json:
```diff
@@ -330,7 +423,7 @@
       "get": {
         "description": "__Windows to IANA Time Zone__\n\n\n**Applies to**: S0 and S1 pricing tiers.\n\n\nThis API returns a corresponding IANA ID, given a valid Windows Time Zone ID. Multiple IANA IDs may be returned for a single Windows ID. It is possible to narrow these results by adding an optional territory parameter.",
-        "operationId": "GetTimezoneWindowsToIANA",
+        "operationId": "GetIANAMetadata",
```
I don't think this returns metadata... It transforms windows id to one or more iana ids. Any better name?

daniel-rocha replied:
Yes this name is wrong. ConvertWindowsTimezoneToIANA

alextts627 committed:
Updated x-ms-examples to be more friendly

alextts627 committed:
Update Descriptions

alextts627 committed:
Update method name

alextts627 commented:
/azp run

daniel-rocha commented on specification/maps/data-plane/Timezone/preview/1.0/timezone.json:
```diff
@@ -209,6 +212,7 @@
           {
             "name": "query",
+            "x-ms-client-name": "coordinate",
             "in": "query",
```
This should be collectionFormat csv with minItems and maxItems both set to 2 and type set to number with format float

alextts627 replied:
Updated

AlexTsao committed:
Update coordinates with collectionFormat csv

AlexTsao committed:
Modified sunrise, sunset type

tjprescott commented on specification/maps/data-plane/Timezone/preview/1.0/timezone.json:
```diff
@@ -538,7 +551,7 @@
         },
         "hasZone1970Location": {
-          "description": "HasZone1970Location property",
+          "description": "This represents the property which the zone has clocks agreeing since 1970",
```
The wording here is confusing to me.

alextts627 replied:
Hi @tjprescott , I have updated the description. Hopefully this will make better sense to you and users. 
Thanks again for the feedback.

tjprescott reviewed (changes requested):
One comment; otherwise, LGTM

alextts627 committed:
Update wording for hasZone1970Location description

alextts627 committed:
Use LatLongPair as reference

alextts627 committed:
Modify exists definitions

1. Combined definitions with same structures into TimezoneResult, TimezoneId, ReferenceTime.

tjprescott reviewed (changes requested):
A couple final comments.

alextts627 committed:
Update Examples to fix Model Validation

alextts627 committed:
Rename to Timezone to be consistent

alextts627 committed:
Update description for hasZone1970Location

daniel-rocha commented:
https://github.com/Azure/azure-rest-api-specs/blob/247802cd132f34d3bb50c1b36c454644e464a8c8/specification/maps/data-plane/Timezone/preview/1.0/timezone.json#L225

This is incorrect. Format "float" should never be used. Only "double". Coordinates (lat/long) are doubles, not floats.

daniel-rocha reviewed (changes requested):
Many changes still to be made for this to be OK.

alextts627 committed:
Minor updates for name, type

daniel-rocha reviewed (changes requested):
Models are named correctly this time but are not working against the APIs. 

```java
Get Timezone IANA Version
2021-10-11 15:49:48.171 [main] [INFO] com.azure.maps.timezone.implementation.TimezoneClientImpl$TimezoneClientService.getIanaVersion - --> GET https://atlas.microsoft.com/timezone/ianaVersion/json?api-version=REDACTED&subscription-key=REDACTEDTry count: 1(empty body)--> END GET
2021-10-11 15:49:49.415 [reactor-http-nio-2] [INFO] com.azure.maps.timezone.implementation.TimezoneClientImpl$TimezoneClientService.getIanaVersion - <-- 200 https://atlas.microsoft.com/timezone/ianaVersion/json?api-version=REDACTED&subscription-key=REDACTED (1184 ms, 19-byte body)Response body:{"Version":"2021a"}<-- END HTTP
{
  "version" : null
}
Get Timezone Windows to IANA
2021-10-11 15:49:49.437 [main] [INFO] com.azure.maps.timezone.implementation.TimezoneClientImpl$TimezoneClientService.convertWindowsTimezoneToIana - --> GET https://atlas.microsoft.com/timezone/windowsToIana/json?api-version=REDACTED&query=REDACTED&subscription-key=REDACTEDTry count: 1(empty body)--> END GET
2021-10-11 15:49:49.614 [reactor-http-nio-2] [INFO] com.azure.maps.timezone.implementation.TimezoneClientImpl$TimezoneClientService.convertWindowsTimezoneToIana - <-- 200 https://atlas.microsoft.com/timezone/windowsToIana/json?api-version=REDACTED&query=REDACTED&subscription-key=REDACTED (163 ms, 66-byte body)Response body:[{"Id":"Pacific/Apia","IsAlias":false,"HasZone1970Location":true}]<-- END HTTP
[ {
  "id" : null,
  "isAlias" : null,
  "aliasOf" : null,
  "hasZone1970Location" : null,
  "alias" : null
} ]
```

daniel-rocha commented:
@alextts627 If you look at the logs of the function calls above, the models returning from the API have attributes starting with Capital letters "Id", "Version", etc. whereas the names in Swagger start with lower letters. That is causing a mismatch and as a result the models are all prepopulated with empty and null strings. 

Fix: match the case of the response body, alias types to correct case using x-ms-client-name.

alextts627 committed:
Modified property to uppercase to match exmaples

alextts627 committed:
Changed back to using RepresentativePoint

alextts627 commented:
It's working in Python now. I was just saying the 'x-ms-client-name' can't be used in these cases.
